File: dqops/src/main/java/com/dqops/metadata/incidents/IncidentNotificationTargetSpec.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */

package com.dqops.metadata.incidents;

import com.dqops.core.secrets.SecretValueLookupContext;
import com.dqops.core.secrets.SecretValueProvider;
import com.dqops.metadata.basespecs.AbstractSpec;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.id.HierarchyNodeResultVisitor;
import com.dqops.utils.docs.generators.SampleStringsRegistry;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.util.Objects;

/**
 * Configuration of addresses used for new or updated incident's notifications.
 * Specifies the webhook URLs or email addresses where the notification messages are sent.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = false)
public class IncidentNotificationTargetSpec extends AbstractSpec implements Cloneable {
    private static final ChildHierarchyNodeFieldMapImpl<IncidentNotificationTargetSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(AbstractSpec.FIELDS) {
        {
        }
    };

    @JsonPropertyDescription("Notification address(es) where the notification messages describing new incidents are pushed using a HTTP POST request (for webhook address) or an SMTP (for email address). The format of the JSON message is documented in the IncidentNotificationMessage object.")
    private String incidentOpenedAddresses;

    @JsonPropertyDescription("Notification address(es) where the notification messages describing acknowledged messages are pushed using a HTTP POST request (for webhook address) or an SMTP (for email address). The format of the JSON message is documented in the IncidentNotificationMessage object.")
    private String incidentAcknowledgedAddresses;

    @JsonPropertyDescription("Notification address(es) where the notification messages describing resolved messages are pushed using a HTTP POST request (for webhook address) or an SMTP (for email address). The format of the JSON message is documented in the IncidentNotificationMessage object.")
    private String incidentResolvedAddresses;

    @JsonPropertyDescription("Notification address(es) where the notification messages describing muted messages are pushed using a HTTP POST request (for webhook address) or an SMTP (for email address). The format of the JSON message is documented in the IncidentNotificationMessage object.")
    private String incidentMutedAddresses;

    /**
     * Returns the addresses where notifications of new incidents are sent.
     * @return Webhook URLs or email addresses.
     */
    public String getIncidentOpenedAddresses() {
        return incidentOpenedAddresses;
    }

    /**
     * Sets addresses where notifications of new incidents are sent.
     * @param incidentOpenedAddresses Webhook URLs or email addresses.
     */
    public void setIncidentOpenedAddresses(String incidentOpenedAddresses) {
        this.setDirtyIf(!Objects.equals(this.incidentOpenedAddresses, incidentOpenedAddresses));
        this.incidentOpenedAddresses = incidentOpenedAddresses;
    }

    /**
     * Returns addresses where notifications of acknowledged incidents are sent.
     * @return Webhook URLs or email addresses
     */
    public String getIncidentAcknowledgedAddresses() {
        return incidentAcknowledgedAddresses;
    }

    /**
     * Sets addresses where notifications of acknowledged incidents are sent.
     * @param incidentAcknowledgedAddresses Webhook url.
     */
    public void setIncidentAcknowledgedAddresses(String incidentAcknowledgedAddresses) {
        this.setDirtyIf(!Objects.equals(this.incidentAcknowledgedAddresses, incidentAcknowledgedAddresses));
        this.incidentAcknowledgedAddresses = incidentAcknowledgedAddresses;
    }

    /**
     * Returns addresses where notifications of resolved incidents are pushed using a HTTP POST request.
     * @return Webhook URLs or email addresses
     */
    public String getIncidentResolvedAddresses() {
        return incidentResolvedAddresses;
    }

    /**
     * Sets addresses where notifications of resolved incidents are sent.
     * @param incidentResolvedAddresses Webhook URLs or email addresses
     */
    public void setIncidentResolvedAddresses(String incidentResolvedAddresses) {
        this.setDirtyIf(!Objects.equals(this.incidentResolvedAddresses, incidentResolvedAddresses));
        this.incidentResolvedAddresses = incidentResolvedAddresses;
    }

    /**
     * Returns addresses where notifications of muted incidents are pushed using a HTTP POST request.
     * @return Webhook URLs or email addresses
     */
    public String getIncidentMutedAddresses() {
        return incidentMutedAddresses;
    }

    /**
     * Sets addresses where notifications of muted incidents are sent.
     * @param incidentMutedAddresses Webhook URLs or email addresses
     */
    public void setIncidentMutedAddresses(String incidentMutedAddresses) {
        this.setDirtyIf(!Objects.equals(this.incidentMutedAddresses, incidentMutedAddresses));
        this.incidentMutedAddresses = incidentMutedAddresses;
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Calls a visitor (using a visitor design pattern) that returns a result.
     *
     * @param visitor   Visitor instance.
     * @param parameter Additional parameter that will be passed back to the visitor.
     */
    @Override
    public <P, R> R visit(HierarchyNodeResultVisitor<P, R> visitor, P parameter) {
        return visitor.accept(this, parameter);
    }

    /**
     * Creates and returns a copy of this object.
     */
    @Override
    public IncidentNotificationTargetSpec deepClone() {
        IncidentNotificationTargetSpec cloned = (IncidentNotificationTargetSpec) super.deepClone();
        return cloned;
    }

    /**
     * Creates a cloned and expanded version of the objects. All parameters are changed to the values expanded from variables like ${ENV_VAR}.
     * @param secretValueProvider Secret value provider.
     * @param lookupContext Secret lookup context.
     * @return Cloned and expanded copy of the object.
     */
    public IncidentNotificationTargetSpec expandAndTrim(SecretValueProvider secretValueProvider, SecretValueLookupContext lookupContext) {
        IncidentNotificationTargetSpec cloned = this.deepClone();
        cloned.incidentOpenedAddresses = secretValueProvider.expandValue(cloned.incidentOpenedAddresses, lookupContext);
        cloned.incidentAcknowledgedAddresses = secretValueProvider.expandValue(cloned.incidentAcknowledgedAddresses, lookupContext);
        cloned.incidentResolvedAddresses = secretValueProvider.expandValue(cloned.incidentResolvedAddresses, lookupContext);
        cloned.incidentMutedAddresses = secretValueProvider.expandValue(cloned.incidentMutedAddresses, lookupContext);
        return cloned;
    }

    /**
     * Combines the incident notification spec with the default addresses. If the incident status' address is null, the corresponding value from default is set.
     * @param defaultIncidentNotification Default incident notification spec.
     * @return Combined IncidentNotificationSpec object with default addresses.
     */
    public IncidentNotificationTargetSpec combineWithDefaults(IncidentNotificationTargetSpec defaultIncidentNotification){
        IncidentNotificationTargetSpec clonedIncidentNotification = this.deepClone();

        if(clonedIncidentNotification.getIncidentOpenedAddresses() == null){
            clonedIncidentNotification.setIncidentOpenedAddresses(defaultIncidentNotification.getIncidentOpenedAddresses());
        }

        if(clonedIncidentNotification.getIncidentAcknowledgedAddresses() == null){
            clonedIncidentNotification.setIncidentAcknowledgedAddresses(defaultIncidentNotification.getIncidentAcknowledgedAddresses());
        }

        if(clonedIncidentNotification.getIncidentResolvedAddresses() == null){
            clonedIncidentNotification.setIncidentResolvedAddresses(defaultIncidentNotification.getIncidentResolvedAddresses());
        }

        if(clonedIncidentNotification.getIncidentMutedAddresses() == null){
            clonedIncidentNotification.setIncidentMutedAddresses(defaultIncidentNotification.getIncidentMutedAddresses());
        }

        return clonedIncidentNotification;
    }

    public static class IncidentNotificationSpecSampleFactory implements SampleValueFactory<IncidentNotificationTargetSpec> {
        @Override
        public IncidentNotificationTargetSpec createSample() {
            return new IncidentNotificationTargetSpec() {{
                setIncidentOpenedAddresses(SampleStringsRegistry.getSampleUrl() + "/opened");
                setIncidentAcknowledgedAddresses(SampleStringsRegistry.getSampleUrl() + "/acknowledged");
                setIncidentResolvedAddresses(SampleStringsRegistry.getSampleUrl() + "/resolved");
                setIncidentMutedAddresses(SampleStringsRegistry.getSampleUrl() + "/muted");
            }};
        }
    }
}

File: distribution/python/dqops/client/models/incident_notification_target_spec.py
from typing import Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

T = TypeVar("T", bound="IncidentNotificationTargetSpec")


@_attrs_define
class IncidentNotificationTargetSpec:
    """
    Attributes:
        incident_opened_addresses (Union[Unset, str]): Notification address(es) where the notification messages
            describing new incidents are pushed using a HTTP POST request (for webhook address) or an SMTP (for email
            address). The format of the JSON message is documented in the IncidentNotificationMessage object.
        incident_acknowledged_addresses (Union[Unset, str]): Notification address(es) where the notification messages
            describing acknowledged messages are pushed using a HTTP POST request (for webhook address) or an SMTP (for
            email address). The format of the JSON message is documented in the IncidentNotificationMessage object.
        incident_resolved_addresses (Union[Unset, str]): Notification address(es) where the notification messages
            describing resolved messages are pushed using a HTTP POST request (for webhook address) or an SMTP (for email
            address). The format of the JSON message is documented in the IncidentNotificationMessage object.
        incident_muted_addresses (Union[Unset, str]): Notification address(es) where the notification messages
            describing muted messages are pushed using a HTTP POST request (for webhook address) or an SMTP (for email
            address). The format of the JSON message is documented in the IncidentNotificationMessage object.
    """

    incident_opened_addresses: Union[Unset, str] = UNSET
    incident_acknowledged_addresses: Union[Unset, str] = UNSET
    incident_resolved_addresses: Union[Unset, str] = UNSET
    incident_muted_addresses: Union[Unset, str] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        incident_opened_addresses = self.incident_opened_addresses
        incident_acknowledged_addresses = self.incident_acknowledged_addresses
        incident_resolved_addresses = self.incident_resolved_addresses
        incident_muted_addresses = self.incident_muted_addresses

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if incident_opened_addresses is not UNSET:
            field_dict["incident_opened_addresses"] = incident_opened_addresses
        if incident_acknowledged_addresses is not UNSET:
            field_dict["incident_acknowledged_addresses"] = (
                incident_acknowledged_addresses
            )
        if incident_resolved_addresses is not UNSET:
            field_dict["incident_resolved_addresses"] = incident_resolved_addresses
        if incident_muted_addresses is not UNSET:
            field_dict["incident_muted_addresses"] = incident_muted_addresses

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        incident_opened_addresses = d.pop("incident_opened_addresses", UNSET)

        incident_acknowledged_addresses = d.pop(
            "incident_acknowledged_addresses", UNSET
        )

        incident_resolved_addresses = d.pop("incident_resolved_addresses", UNSET)

        incident_muted_addresses = d.pop("incident_muted_addresses", UNSET)

        incident_notification_target_spec = cls(
            incident_opened_addresses=incident_opened_addresses,
            incident_acknowledged_addresses=incident_acknowledged_addresses,
            incident_resolved_addresses=incident_resolved_addresses,
            incident_muted_addresses=incident_muted_addresses,
        )

        incident_notification_target_spec.additional_properties = d
        return incident_notification_target_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
